kconfig: add support for warnings when using experimental features

tejlmand opened the pull request:
This PR allows a user to have a warning printed if any experimental features are used in Zephyr.

As example, building `samples/hello_world` with `THREAD_MONITOR` which is experimental will look as:
```
$ cmake -GNinja -DBOARD=nrf52840dk_nrf52840 -Bbuild samples/hello_world/ -DCONFIG_THREAD_MONITOR=y
Parsing /projects/github/ncs/zephyr/Kconfig
Loaded configuration '/projects/github/ncs/zephyr/boards/arm/nrf52840dk_nrf52840/nrf52840dk_nrf52840_defconfig'
Merged configuration '/projects/github/ncs/zephyr/samples/hello_world/prj.conf'
Merged configuration '/projects/github/ncs/zephyr/build/zephyr/misc/generated/extra_kconfig_options.conf'
Configuration saved to '/projects/github/ncs/zephyr/build/zephyr/.config'
Kconfig header saved to '/projects/github/ncs/zephyr/build/zephyr/include/generated/autoconf.h'
-- The C compiler identification is GNU 7.3.1
....
```

Enabling this new warning together with `THREAD_MONITOR`, will give:
```
$ cmake -GNinja -DBOARD=nrf52840dk_nrf52840 -Bbuild samples/hello_world/ -DCONFIG_WARN_EXPERIMENTAL=y -DCONFIG_THREAD_MONITOR=y
Parsing /projects/github/ncs/zephyr/Kconfig
Loaded configuration '/projects/github/ncs/zephyr/boards/arm/nrf52840dk_nrf52840/nrf52840dk_nrf52840_defconfig'
Merged configuration '/projects/github/ncs/zephyr/samples/hello_world/prj.conf'
Merged configuration '/projects/github/ncs/zephyr/build/zephyr/misc/generated/extra_kconfig_options.conf'
Configuration saved to '/projects/github/ncs/zephyr/build/zephyr/.config'
Kconfig header saved to '/projects/github/ncs/zephyr/build/zephyr/include/generated/autoconf.h'

warning: Experimental symbol THREAD_MONITOR is enabled.

.....
```

--------

This adds a new setting which allows a user to print a warning when
a feature that is marked `[EXPERIMENTAL]` is enabled.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
kconfig: add support for warnings when using experimental features

This adds two new Kconfig settings.

The first setting `EXPERIMENTAL` which is a promptless symbol.
This symbol must be selected by any setting which is used to enable
an experimental feature.

The second setting is `WARN_EXPERIMENTAL` which is a user controlled
setting that configures the build system to print a warning when
experimental features are enabled.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

galak reviewed (dismissed):
I like the idea, but I don't think we should be modifying scripts/kconfig/kconfig.py to do this.  Maybe a separate script similar to scripts/kconfig/hardenconfig.py?

tejlmand commented:
> I like the idea, but I don't think we should be modifying scripts/kconfig/kconfig.py to do this. Maybe a separate script similar to scripts/kconfig/hardenconfig.py?

Except that `hardenconfig.py` is a build target that users can call with `ninja hardenconfig`, similar to `menuconfig` and other extra targets, and thus only take resources when it's called.

The `CONFIG_WARN_EXPERIMENTAL=y` allows you to get a warning if experimental features are enabled as part of the CMake invocation, similar to other build warnings.

The motivation for updating `kconfig.py` is because it has the parsed kconfig tree available, and thus no overhead of re-invoking Kconfig parsing at CMake time.

tejlmand commented:
@galak also notice, `kconfig.py` already does various checking of the Kconfig tree today:
https://github.com/zephyrproject-rtos/zephyr/blob/16d2ab790ce074889757185e7ac21c2618c80ef5/scripts/kconfig/kconfig.py#L57

https://github.com/zephyrproject-rtos/zephyr/blob/16d2ab790ce074889757185e7ac21c2618c80ef5/scripts/kconfig/kconfig.py#L62

https://github.com/zephyrproject-rtos/zephyr/blob/16d2ab790ce074889757185e7ac21c2618c80ef5/scripts/kconfig/kconfig.py#L63

rugeGerritsen commented:
In a very limited set of places we have constructs like:
```
config BT_CTLR_ADV_EXT
	bool "LE Advertising Extensions" if !BT_LL_SW_SPLIT

config BT_CTLR_ADV_EXT
	bool "LE Advertising Extensions [EXPERIMENTAL]" if BT_LL_SW_SPLIT
```
I would assume that we need to rewrite those to something similar to
```
config BT_CTLR_ADV_EXT
	bool "LE Advertising Extensions"

config BT_CTLR_ADV_EXT_SW_LL_SPLIT
	bool "LE Advertising Extensions [EXPERIMENTAL]"
	default y if BT_LL_SW_SPLIT && BT_CTLR_ADV_EXT
```
Otherwise `warning: Experimental symbol BT_CTLR_ADV_EXT` will be printed even if `BT_LL_SW_SPLIT` is not selected. See https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/bluetooth/controller/Kconfig#L460

Would it be possible to make a similar construct that fails the configuration, as opposed to printing a warning, if an experimental feature is selected?

tejlmand commented:
> I would assume that we need to rewrite those to something similar to

Let me take a second look.
Right now a warning is printed if symbol is enabled, but I did consider to only print the warning for enabled symbols if they were also visible.
I assume that would help in this particular case.

The disadvantage is of course that a symbol marked `[EXPERIMENTAL]` will not be printed if it is not visible, but I don't know if we have any such cases where an experimental symbol can be enable but not shown, where we would like the warning.

Main point: we need to consider best user experience on this.

> Would it be possible to make a similar construct that fails the configuration, as opposed to printing a warning, if an experimental feature is selected?

It would, but then you should take this issue into consideration #9573
meaning that failing the configuration currently prevents the user from entering menuconfig to fix the reason.

tejlmand commented:
@rugeGerritsen you probably need to take a look at that Kconfig snippet anyway, cause if `BT_LL_SW_SPLIT=n` you'll see this is menuconfig today, which I assume is not your intention:
![image](https://user-images.githubusercontent.com/9802583/112058981-8f749880-8b5b-11eb-8553-db429d5b4650.png)

nashif commented:
we could revert back to the original experimental support in Kconfig where each experimental Kconfig would select EXPERIMENTAL and then just print something out if any experimental features are enabled....

tejlmand commented:
> we could revert back to the original experimental support in Kconfig where each experimental Kconfig would select EXPERIMENTAL and then just print something out if any experimental features are enabled....

if we did have something like this back in time, then it might be worth knowing why it was removed.
Did users forget to make a `select EXPERIMENTAL` or was it insufficient to just print experimental in use, and not know which feature that was the root cause ?

With this PR, users would be informed about the symbol
> warning: Experimental symbol THREAD_MONITOR is enabled.

tejlmand commented:
@rugeGerritsen only printing if prompt is visible, but to test with your example you would need to update the Kconfig.

galak commented:
> @galak also notice, `kconfig.py` already does various checking of the Kconfig tree today:

Those checks don't seem to be zephyr specific, which is my concern about adding this check directly to scripts/kconfig/kconfig.py

tejlmand commented:
> 
> 
> > @galak also notice, `kconfig.py` already does various checking of the Kconfig tree today:
> 
> Those checks don't seem to be zephyr specific, which is my concern about adding this check directly to scripts/kconfig/kconfig.py

Those checks are exactly Zephyr specific.
They are needed because Zephyr samples / tests uses handwritten Kconfig files in the form of `prj.conf`.
Because of this, in kconfig.cmake, we do this:
https://github.com/zephyrproject-rtos/zephyr/blob/c1f7b9f45a78d7e082a131474cf5c4592a1aaba8/cmake/kconfig.cmake#L239-L240

when we generate the first `.config` file.
In that case we want to carry out the previous 3 checks I linked.

Once we have the `.config` in place, even promptless symbols gets values assigned.

As example, kconfiglib itself allows value assignment to promptless symbols, but in Zephyr we don't want values assigned to promptless symbols in `prj.conf` and other Kconfig fragments. That's why we have those checks in kconfig.py, which is a Zephyr specific python script.

carlescufi commented:
> > @galak also notice, `kconfig.py` already does various checking of the Kconfig tree today:
> 
> Those checks don't seem to be zephyr specific, which is my concern about adding this check directly to scripts/kconfig/kconfig.py

As @tejlmand pointed out, they are very much zephyr-specific. `kconfig.py` is a wrapper for `kconfiglib.py` that handles all the zephyr-specific behavior.

pabigot commented:
> if we did have something like this back in time, then it might be worth knowing why it was removed.

f37a4fcd71e09fde122657beddc5ee6e6b94c65e  appears to have been part of a batch of Kconfig cleanups before Zephyr's initial release.  I see no identified rationale.

tejlmand commented:
>     * Rationale for the removal of `depend on EXPERIMENTAL` from the Linux kernel: https://lwn.net/Articles/520867/

Thanks for the link.
That link has a very good point I think we should adopt:
>   For items that really are experimental, maintainers should use "default
  n", optionally include "(EXPERIMENTAL)" in the title, and add language to
  the help text indicating why the item should be considered experimental.

I think it would be good if it was clearer why a feature is considered experimental.
If others agree, I will open an issue for followup on improving help texts.

tejlmand commented:
marked `DNM` until https://github.com/zephyrproject-rtos/zephyr/pull/33566#issuecomment-810396251 has been completed.

rugeGerritsen commented:
@tejlmand , will this issue be reopened?

carlescufi commented:
> @tejlmand , will this issue be reopened?

Reopened now

tejlmand committed:
kconfig: bluetooth: experimental settings now uses select EXPERIMENTAL

With the introduction of `EXPERIMENTAL` and `WARN_EXPERIMENTAL` in
Zephyr all subsys/bluetooth and drivers/bluetooth/hci settings having
`[EXPERIMENTAL]` in their prompt has has been updated to include
`select EXPERIMENTAL` so that developers can enable warnings when
experimental features are enabled.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
kconfig: canbus: experimental settings now uses select EXPERIMENTAL

With the introduction of `EXPERIMENTAL` and `WARN_EXPERIMENTAL` in
Zephyr all subsys/canbus, subsys/net/l2/canbus, and drivers/can settings
having `[EXPERIMENTAL]` in their prompt has has been updated to include
`select EXPERIMENTAL` so that developers can enable warnings when
experimental features are enabled.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

carlescufi reviewed (approved):
This is now as per agreed during the conversations.

nashif reviewed (changes requested):
1. need to see some documentation about this and how this can be used. We do not want people depending on this kconfig and this is not documented anywhere, this can have negative results.
2. I remember seeing a change that had all experimental features converted to use this, we need to have at least few places use this, consider it a test and a reference to be added in the docs.

tejlmand committed:
doc: application: added description of WARN_EXPERIMENTAL setting

This commit adds description of the `WARN_EXPERIMENTAL` Kconfig setting
to the application development guide.

It describes the use of `CONFIG_WARN_EXPERIMENTAL` along with an
example of the warning output when experimental features are enabled.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>
